Add support for multicast IPv4 TTL and IPv6 hop limit

jukkar opened the pull request:
Add option support for adjusting the IPv4 multicast time-to-live value. Add also tests for both unicast and multicast TTL settings.
Add also similar IPv6 hop limit option for both unicast and multicast IPv6 packets + tests.
Add checks and tests for IPv4 TTL 0 or IPv6 hop limit 0 when sending packet, the packet should dropped in this case.

Fixes #60299

jukkar committed:
net: context: Add support for adjusting IPv4 multicast ttl

Add option support for adjusting the IPv4 multicast
time-to-live value.

Fixes #60299

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: net: udp: Add IPv4 time-to-live multicast tests

Make sure that setting IPv4 multicast TTL works as expected.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: Add support for adjusting IPv6 multicast hop limit

Add option support for adjusting the IPv6 multicast
multicast hop limit value.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: net: udp: Add IPv6 multicast hop limit tests

Make sure that setting IPv6 multicast hop limit works as expected.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: if: Fix typo in IPv6 hop limit API name

The net_if_ipv6_set_hop_limit() API was missing the "_if_"
part in it. Fix this so that the network interface API is
consistent. The old function is deprecated and should not
be used. The old function is left to the code and it calls
the new properly named function.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: Add support for adjusting IPv6 unicast hop limit

Add option support for adjusting the IPv6 unicast
hop limit value.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: net: udp: Add IPv6 unicast hop limit tests

Make sure that setting IPv6 unicast hop limit works as expected.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

pdgendt commented on subsys/net/ip/Kconfig.ipv4:
```diff
@@ -17,6 +17,12 @@ config NET_INITIAL_TTL
 	help
 	  The value should be > 0
 
+config NET_INITIAL_MCAST_TTL
+	int "Initial time to live for multicast packets"
+	default NET_INITIAL_TTL
+	help
+	  The value should be > 0
```
You should add a `range` (same for `NET_INITIAL_TTL`)

pdgendt commented on subsys/net/ip/Kconfig.ipv4:
```diff
@@ -14,6 +14,14 @@ if NET_IPV4
 config NET_INITIAL_TTL
 	int "Initial time to live for a connection"
 	default 64
+	range 0 255
+	help
+	  The value should be > 0
```
Why allow the value to be set to `0` if the description states the opposite?

jukkar replied:
0 is a valid value but not very useful usually so that is why it says "should". We actually should drop the packet if ttl is 0 (in TX), I just realized that we do not do that, I will send a patch for that. For RX, we drop the packet already if ttl of incoming packet is 0.

pdgendt replied:
I'm sorry, the description is confusing (it was before this change of course), and could use some additional context.

jukkar replied:
Hmm, I could try to add more information there although I am not sure how much about IP protocol functionality should be described here. Anyway, I will figure out something.

jukkar replied:
@pdgendt Would this description be helpful
```
	  The value should normally be > 0. The device receiving the IPv4
	  packet will decrement the value and will drop the packet if the TTL
	  value is 0. When sending, the packet is dropped before transmitting
	  to network if TTL is 0.
```

pdgendt replied:
LGTM

pdgendt commented on include/zephyr/net/net_if.h:
```diff
@@ -1672,7 +1678,25 @@ uint8_t net_if_ipv6_get_hop_limit(struct net_if *iface);
  * @param hop_limit New hop limit
  */
-void net_ipv6_set_hop_limit(struct net_if *iface, uint8_t hop_limit);
+void net_if_ipv6_set_hop_limit(struct net_if *iface, uint8_t hop_limit);
```
This is a public API change, I think you should leave `net_ipv6_set_hop_limit` in, simply calling `net_if_ipv6_set_hop_limit` and add a deprecation warning.

jukkar replied:
Good point, will fix.

jukkar committed:
net: Refactor IP checks just before sending packets

* Check IPv4 TTL or IPv6 hop limit and drop the packet if
  the value is 0
* Check the IP addresses so that we do the loopback check
  at runtime if the packet is destined to loopback interface.
* Update the statistics properly for dropped packets.
* Do not update sent packets if we drop packets.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: net: udp: Add IPv4 TTL 0 test

Make sure packet is dropped if TTL 0 packet is tried to
send.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: socket: Add support for adjusting IPv4 TTL

The IPv4 TTL could only manipulated via net_context interface.
It makes sense to allow the same from socket interface via
the setsockopt/getsockopt calls.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: net: udp: Add IPv4 unicast TTL tests

Make sure that setting IPv4 unicast TTL works as expected.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: net: udp: Verify that packet is dropped from stats

Make sure that statistics is properly update for dropped
packets when IPv4 TTL is 0 or IPv6 hop limit is 0.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: net: udp: Add IPv6 hop limit 0 test

Make sure packet is dropped if IPv6 hop limit 0 packet is
tried to send.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
As a general note, I originally planned to implement the multicast TTL/hoplimit support only, but noticed that the IPv4 TTL and IPv6 hoplimit were a bit flaky, so the changes are more extensive. We were also missing tests for unicast versions, those + multicast ones are now implemented.

pdgendt commented on subsys/net/ip/net_context.c:
```diff
@@ -1544,6 +1549,86 @@ static int get_context_dscp_ecn(struct net_context *context,
+static int get_context_ttl(struct net_context *context,
+				 void *value, size_t *len)
+{
+#if defined(CONFIG_NET_IPV4)
+	*((int *)value) = context->ipv4_ttl;
+
+	if (len) {
+		*len = sizeof(int);
+	}
+
+	return 0;
+#else
```
Did you not introduce generic getter/setter function to be used here?

jukkar replied:
For the ttl and hoplimit getters I did not, the `get_uint8_option()` expects user to supply uint8_t value but the value for the ttl and hoplimit is typically int, so the values were incompatible. Then I thought about creating `get_int_option()` but because we are storing ttl and hoplimit as uint8_t the function was not that nice looking. I will investigate this option getter/setter dilemma in separate PR that I mentioned in the refactoring patch.

jukkar committed:
net: dns: Do not pass 0 as TTL or hop limit

We specifically set TTL/hoplimit to 1 for LLMNR,
but only want to set it if in that specific case.
We must not pass TTL/hoplimit value 0 as that would
cause the packet to be dropped.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: dns: responders: Set the multicast TTL or hoplimit

We are creating a multicast address in mDNS or LLMNR
responder so set the TTL or hoplimit using the multicast
variant API.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: net: dns-sd: Fix IPv4 TTL and IPv6 hop limit checking

We need to use multicast TTL/hoplimit instead of unicast one
in the test.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

rlubos commented on subsys/net/ip/Kconfig.ipv4:
```diff
@@ -12,10 +12,27 @@ menuconfig NET_IPV4
+	  to network if TTL is 0.
+
+config NET_INITIAL_MCAST_TTL
+	int "Initial IPv4 time to live for multicast packets"
```
`time to live value` to align with the change above?

rlubos commented on subsys/net/ip/Kconfig.ipv4:
```diff
@@ -12,10 +12,27 @@ menuconfig NET_IPV4
 
 config NET_INITIAL_TTL
-	int "Initial time to live for a connection"
+	int "Initial IPv4 time to live value for a connection"
```
`for a connection` sounds a bit awkward in IP context, maybe `for unicast packets` similar as below, so it's explicit that the config is now only about unicast?

rlubos commented on subsys/net/ip/Kconfig.ipv6:
```diff
@@ -40,10 +40,27 @@ config NET_IF_IPV6_PREFIX_COUNT
 
 config NET_INITIAL_HOP_LIMIT
-	int "Initial hop limit for a connection"
+	int "Initial IPv6 hop limit for a connection"
```
Similar comment as for IPv4, shall we mention unicast packets explicitly?

rlubos reviewed (dismissed):
Just some doc suggestions to consider, but overall looks good. Shall we mention the API changes in the changelog?

jukkar committed:
doc: migration-guide: 3.6: IPv4 TTL and IPv6 hoplimit changes

Add information about IPv4 TTL and IPv6 hoplimit changes
for unicast and multicast packets.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
Minor tweaks to kconfig file. Added also info to migration guide.

kartben commented on doc/releases/migration-guide-3.6.rst:
```diff
@@ -157,6 +157,11 @@ Networking
   a NULL pointer. IGMPv3 can be enabled using the Kconfig ``CONFIG_NET_IPV4_IGMPV3``.
   (:github:`65293`)
 
+* The network stack uses now separate IPv4 TTL (time-to-live) value for multicast packets. Before
+  the same TTL value was used for unicast and multicast packets. The IPv6 hop limit value is also
+  changed so that unicast and multicast packets can have different hop limit values.
+  (:github:`65886`)
```
Some grammar fixes below, but wondering if this is not "just" something for the release notes? Are people's existing apps expected to *break* as per the described change, or is it just that from now on people have more flexibility w.r.t. configuring TTLs?

It might also make sense to mention the new Kconfig options (but again, might be a "release note" thing rather than migration guide, idk)

```suggestion
* The network stack now uses separate IPv4 TTL (time-to-live) values for multicast packets. Before,
  the same TTL value was used for unicast and multicast packets. The IPv6 hop limit value is also
  changed so that unicast and multicast packets can have different hop limit values.
  (:github:`65886`)
```

kartben replied:
> > The network stack now uses separate IPv4 TTL (time-to-live) values for multicast packets. Before,
> 
> @kartben Hmm, about the plural "values" that sounds wrong as there is just one "value". Each application can have their own value so perhaps "values" could be used but it kind of feels wrong. WDYT?

ah! you're right :) I fixed the wrong end of the sentence I guess, just updated the suggested change. and ACK re: waiting for others like @rlubos to chime in, you guys def know better in any case :)

jukkar replied:
I slightly modified the IPv6 hop limit sentence, does this look ok now?
```
* The network stack now uses a separate IPv4 TTL (time-to-live) value for multicast packets.
  Before, the same TTL value was used for unicast and multicast packets.
  The IPv6 hop limit value is also changed so that unicast and multicast packets can have a
  different one. (:github:`65886`)
```

jukkar committed:
doc: release: 3.6: IPv4 TTL and IPv6 hoplimit changes

Add information about IPv4 TTL and IPv6 hoplimit changes
for unicast and multicast packets.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
Release and migration note changes in this version.

kartben commented on include/zephyr/net/net_context.h:
```diff
@@ -703,6 +709,17 @@ static inline void net_context_set_ipv4_ttl(struct net_context *context,
 	context->ipv4_ttl = ttl;
 }
 
+static inline uint8_t net_context_get_ipv4_mcast_ttl(struct net_context *context)
```
there are quite a few of these `net_context_get|set_*` functions (existing ones, and the ones you added) that are missing doxygen comments. Could you please add them? Should these be considered internal (I have no idea myself :)) you may consider not having the doxygen documentation (although it's always nice to have imo), and simply enclose them in ` @cond INTERNAL_HIDDEN` sections.

Sorry that I didn't notice earlier as I really just paid attention to the migration guide changes when the bot added me.

kartben commented on include/zephyr/net/net_context.h:
```diff
@@ -348,8 +348,14 @@ __net_socket struct net_context {
 	union {
-		uint8_t ipv6_hop_limit;
-		uint8_t ipv4_ttl;
+		struct {
+			uint8_t ipv6_hop_limit;
+			uint8_t ipv6_mcast_hop_limit;
+		};
+		struct {
+			uint8_t ipv4_ttl;
+			uint8_t ipv4_mcast_ttl;
+		};
```
would be nice to have doxygen comments here as well (again, unless this is something you want to hide from the public API).

jukkar replied:
Sure, I will submit a follow up pr adding more comments. Thx!

kartben reviewed (commented):
Just noticed missing Doxygen comments -- not making this blocking as I don't necessarily want to slow you guys down as I trust you'll address this in a follow-up PR, but would still be nice to have I think, especially since the networking headers are usually pretty well documented, so would be good to keep the track record ;-)

https://builds.zephyrproject.io/zephyr/pr/65886/api-coverage/zephyr/net/index.html
